SNIPS/snip-9.md

```diff
@@ -0,0 +1,108 @@
+```
+
+- **caller**: can be used to restrict which calling contracts can initiate this execution, although a special address `'ANY_CALLER'` can be used to allow every caller.
+- **nonce**: this is different from the account’s usual nonce, it is used to prevent signature reuse across executions and doesn’t need to be incremental as long as it’s unique.
+- **execute_{before,after}**: timestamp range in which the execution is allowed.
+- **calls**: the usual calls to be executed by the account.
+
+### 2. Sign it using ERC-712 typed data hashing
+
+Reference implementation: [https://github.com/argentlabs/argent-contracts-starknet/blob/main/tests/lib/outsideExecution.ts](https://github.com/argentlabs/argent-contracts-starknet/blob/main/tests/lib/outsideExecution.ts)
+
+See this SNIP for more info on offchain signatures on Starknet: [https://community.starknet.io/t/snip-off-chain-signatures-a-la-eip712/98029](https://community.starknet.io/t/snip-off-chain-signatures-a-la-eip712/98029)
+
+### 3. Pass the struct and signature to the account
+
+Check if the account supports this SNIP:
+
+```rust
+let acccount = IErc165Dispatcher { contract_address: acount_address };
+let is_supported = account.supports_interface(ERC165_OUTSIDE_EXECUTION_INTERFACE_ID); // see below for actual value
+```
+
+Call the `execute_from_outside` method on the account:
+
+```rust
+let acccount = IOutsideExecutionDispatcher { contract_address: acount_address };
+// pre-execution logic...
+let results = account.execute_from_outside(outside_execution, signature);
+// post-execution logic...
+```
+
+### For account buildoors
+
+To accept such outside transactions, the account contract must implement the following interface:
+
+```rust
+/// Interface ID: 0x68cfd18b92d1907b8ba3cc324900277f5a3622099431ea85dd8089255e4181
+#[starknet::interface]
+trait IOutsideExecution<TContractState> {
+    /// @notice This method allows anyone to submit a transaction on behalf of the account as long as they have the relevant signatures
+    /// @param outside_execution The parameters of the transaction to execute
+    /// @param signature A valid signature on the ERC-712 message encoding of `outside_execution`
+    /// @notice This method allows reentrancy. A call to `__execute__` or `execute_from_outside` can trigger another nested transaction to `execute_from_outside`.
+    fn execute_from_outside(
+        ref self: TContractState,
+        outside_execution: OutsideExecution,
+        signature: Array<felt252>,
+    ) -> Array<Span<felt252>>;
+
+    /// Get the status of a given nonce, true if the nonce is available to use
+    fn is_valid_outside_execution_nonce(
+        self: @TContractState,
+        nonce: felt252
+    ) -> bool;
+}
+```
+
+Reference implementation: [https://github.com/argentlabs/argent-contracts-starknet/blob/main/contracts/account/src/argent_account.cairo#L247](https://github.com/argentlabs/argent-contracts-starknet/blob/main/contracts/account/src/argent_account.cairo#L247)
```

Comment:
Should not have a reference implementation, for neutrality concerns.

Reply:
I see a lot of EIPs having a reference implementation, I don't see the issue.
Actually, as a dev I'd rather have an implementation example for such use cases. 
It'll allow me to quickly jump in and develop my own.

Only concern is that it can get out of date.

Reply:
i'll remove the link to the argent repo and replace it with some inline sample code

Reply:
done